### PR TITLE
Externalize query cache into its own service

### LINKED ISSUE #1440: Allow externalized Query Cache
This is a ticket to track the the separation of the JPA server query cache into its own service, so that it can be left intact or subbed out with an alternate implementation.

The initial alternate implementation will be the InfiniSpan cache.

- [x] Split the query cache into its own service. This service should have the exact same semantics as the existing implementation, so it is just a refactoring of existing DAO code out of the DAO.
- [x] Bonus: Make sure that we don't hammer the DB with updates to the "Search last used" time if a seach is reused frequently (we do this currently)

### PR DESCRIPTION
**Pull request by jamesagnew**
Fixes #1440

### CODE REVIEW

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/SearchCoordinatorSvcImpl.java [hunk 1]
```diff
@@ -451,12 +303,152 @@ public IBundleProvider registerSearch(final IDao theCallingDao, final SearchPara
...
+			includedPids.addAll(theSb.loadIncludes(myContext, myEntityManager, pids, theParams.getIncludes(), false, theParams.getLastUpdated(), "(synchronous)", theRequestDetails));
+			List<Long> includedPidsList = new ArrayList<>(includedPids);
+
+			List<IBaseResource> resources = new ArrayList<>();
+			theSb.loadResourcesByPid(pids, includedPidsList, resources, false, theRequestDetails);
+			return new SimpleBundleProvider(resources);
+		});
+	}
+
+	@org.jetbrains.annotations.Nullable
```
Comment (jamesagnew) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/SearchCoordinatorSvcImpl.java]: Wrong nullable

Comment (fil512) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/SearchCoordinatorSvcImpl.java]: I changed it, I hope to the right one.

**Comment by fil512** [comment]
@jamesagnew this is now ready for review

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/SearchCoordinatorSvcImpl.java [hunk 2]
```diff
@@ -79,6 +78,7 @@
 import javax.annotation.Nullable;
 import javax.annotation.PostConstruct;
 import javax.persistence.EntityManager;
+import javax.validation.constraints.NotNull;
```
Comment (jamesagnew) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/SearchCoordinatorSvcImpl.java]: Haha wrong one... Should be `Nonnull`

**Comment by jamesagnew** [comment]
Approved

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/SearchCoordinatorSvcImpl.java (hunk 2 above)
Comment (fil512) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/SearchCoordinatorSvcImpl.java]: What’s the full path?

Comment (jamesagnew) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/SearchCoordinatorSvcImpl.java]: `javax.annotation.Nonnull`

**Comment by fil512** [comment]
He javadoc for that annotation says “This annotation is rarely written in
source code, because it is the default“ which suggests there is no
advantage to adding it,

**Pull request merged**